### PR TITLE
Add dimension, metric and namespace validation

### PR DESCRIPTION
**Pull request by markkuhn**
Related to:
* https://github.com/awslabs/aws-embedded-metrics-node/pull/135
* https://github.com/awslabs/aws-embedded-metrics-dotnet/pull/37
* https://github.com/awslabs/aws-embedded-metrics-java/pull/119

*Description of changes:*
Added validation for dimension, metric and namespace

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_embedded_metrics/validator.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            DimensionSetExceededError: If the dimension set is too large
+            InvalidDimensionError: If a dimension is invalid
+    """
+    if len(dimension_set) > constants.MAX_DIMENSION_SET_SIZE:
+        raise DimensionSetExceededError(
+            f"Maximum number of dimensions per dimension set allowed are {constants.MAX_DIMENSION_SET_SIZE}")
+
+    for key, value in dimension_set.items():
+        if not key or len(key.strip()) == 0:
+            raise InvalidDimensionError("Dimension key must include at least one non-whitespace character")
```
Comment (gordonpn) [inline aws_embedded_metrics/validator.py:39]: nit: we call them `dimension name` in the java library and in the docs. I think it should be `dimension name` for consistency.

same with line 45 below

File: tests/logger/test_metrics_context.py [hunk 2]
```diff
@@ -258,6 +287,28 @@ def test_put_metric_uses_none_unit_if_not_provided():
     assert metric.unit == "None"
 
 
+@pytest.mark.parametrize(
+    "name, value, unit",
+    [
+        ("", 1, "None"),
```
Comment (gordonpn) [inline tests/logger/test_metrics_context.py:293]: If the user passes `put_metric("Example", 1.0, None)`

(No quotes around `None`).

What happens? That should pass, right?

Comment (markkuhn) [reply tests/logger/test_metrics_context.py:293]: Yes, that will pass. Using `None` as a metric unit will set it to `"None"` in the metric constructor `metric.py`.

Comment (gordonpn) [reply tests/logger/test_metrics_context.py:293]: Awesome

**Review by gordonpn — APPROVED** [review]

**Review by evanuk — APPROVED** [review]

**Comment by hussam789** [comment]
## PR Code Suggestions ✨



<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>Possible issue</td>
<td>



<details><summary>Fix incorrect constant usage</summary>

___


**The metric name length validation is using the wrong constant. It should use <br>MAX_METRIC_NAME_LENGTH instead of MAX_DIMENSION_NAME_LENGTH.**

[aws_embedded_metrics/validator.py [75-76]](https://github.com/awslabs/aws-embedded-metrics-python/pull/91/files#diff-bab0a43bff403f0d35ba2a433e7b2b02cc93dd34194edd292715d66dcaf3a19dR75-R76)

```diff
-if len(name) > constants.MAX_DIMENSION_NAME_LENGTH:
-    raise InvalidMetricError(f"Metric name cannot be longer than {constants.MAX_DIMENSION_NAME_LENGTH} characters")
+if len(name) > constants.MAX_METRIC_NAME_LENGTH:
+    raise InvalidMetricError(f"Metric name cannot be longer than {constants.MAX_METRIC_NAME_LENGTH} characters")
```

- [ ] **Apply this suggestion** 

<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: The suggestion correctly identifies a bug where the metric name length is validated against the wrong constant, and replacing MAX_DIMENSION_NAME_LENGTH with MAX_METRIC_NAME_LENGTH ensures metrics are correctly validated.


</details></details></td><td align=center>High

</td></tr>
<tr><td align="center" colspan="2">

- [ ] More 

</td><td></td></tr></tbody></table>

**Pull request merged**